### PR TITLE
TICKET-6:

### PR DESCRIPTION
**Pull request by simyara**
related to ticket [Ticket-6](https://trello.com/c/gbpCw3Ym/4-06-add-remove-and-create-api-method)
pull request updated

### CODE REVIEW

File: actions/singlePicture.js [hunk 1]
```diff
@@ -69,5 +77,96 @@ module.exports = {
...
+            name: {type: 'string'},
+            details: {
+                type :{
+                    url: {type: 'string'},
+                    description: {type: 'string'}
+                }
+            }
+        };
+
+        let validationResult = yield validator.validateObject(this.request.body, schema);
```
Comment (ikryvanos) [inline actions/singlePicture.js]: Why you need to validate such schema before main one?

File: actions/singlePicture.js [hunk 2]
```diff
@@ -69,5 +77,96 @@ module.exports = {
...
+                    error: validationReqResult.errorMessage
+                }
+            };
+            return;
+        }
+
+        let pictureData = yield pictureServer.findOne(this.params.id, this.request.body);
+
+        if (pictureData) {
+            this.status = 404;
```
Comment (ikryvanos) [inline actions/singlePicture.js]: 404 status code not suitable for such case

File: services/pictureServer.js [hunk 3]
```diff
@@ -88,7 +91,15 @@ module.exports = {
...
+    },
+    *putOne(id, item) {
+        let newItem = lodash.merge({id: id}, item);
+        picItems.push(newItem);
+        return newItem;
+    },
+    *deleteOne(id){
+        picItems = picItems.filter((item) => item.id !== id);
+        return id;
+        }
```
Comment (ikryvanos) [inline services/pictureServer.js]: Wrong tabulation

File: services/validator.js [hunk 4]
```diff
@@ -34,6 +34,36 @@ function validate(obj, schema) {
     return resObj;
 }
 
+//all fields in schema marked as required must be presented in object
+function validateRequired(obj, schema) {
```
Comment (ikryvanos) [inline services/validator.js]: Make only one method for validation

**Review by ikryvanos — CHANGES_REQUESTED** [review]

File: services/validator.js [hunk 5]
```diff
@@ -1,52 +1,38 @@
...
         }
 
-        if ((typeof obj[element]) !== schema[element].type) {
-            errorMessage = element + ' type mismatch';
+        if (schema[key].type === 'object') {
+            if ((typeof obj[key]) === 'object') {
+                let res = validateObject(obj[key], schema[key].properties);
+                return res.errorMessage;  //(*) recursion
+            }
+            errorMessage = key + ' type mismatch';
```
Comment (satamail) [inline services/validator.js]: Try to avoid error message duplication

**Review by ikryvanos — APPROVED** [review]

**Pull request merged**